### DISCUSSION TITLE
Store mp4 file during publish to stream process

### DISCUSSION THREAD

**Discussion by zt1983811** [bluenviron/mediamtx#991] [Questions]
## Context:
We would like try to store mp4 video file during the stream publish (on protocol RTMP and optionally RTSP in future) process instead of publish then stream to store. We understand there is a way to store video by re-streaming from RTMP protocol like this 
![image](https://user-images.githubusercontent.com/2559543/173600051-98b0e7c3-1f73-40e7-9110-3bddd543851b.png)
However, we don't want publish then stream save, and we would like publish and save in one process

### Approach 1 
save in `stream.go`  function `writeData(data *data)`  , convert RTP stream to mp4 then save
#### Reason choose here:
It seems all protocol will call this function to write data to PacketRTP for streaming
#### Problem:
The problem we have is the data here is already encoded with h264, we are not sure how to decode and convert the format to mp4 format. Also we are not sure if it will be efficient to do decode and format conversion here

### Approach 2
reuse `muxer.go` function `File(name string, msn string, part string, skip string)`
#### Reason choose here:
In the File function if the Muxer variant is mp4 type, it will call `muxer_variant_fmp4.go`  to init mp4 file we believe. However, we are new to this and not exactly sure what really is.
#### Problem:
We create `NewMuxer` in `rtmp_conn.go`  function `runPublish` right after we get videoTrack and audioTrack see screenshot below. However when I check the file I saved in disk it's a file only 2 kb and can't open by mp4 player, it complains on the file format is not correct.
![image](https://user-images.githubusercontent.com/2559543/173604992-6b7f7631-0092-4b4f-9953-43202738029e.png)

## Question:
* Do you think if we are in the right approach?
  * if it's which one do you think it makes more sense? 
  * if not could you suggest something else?
* If we need to change anything in stream.go or rtmp_conn.go, instead directly change in file, is there any easier way to change the behavior of the file?

**Comment by ilterpehlivan** [comment]
Hi @aler9 ,
we also have a requirement to save data into different clouds like Azure, AWS etc. with readable format (MP4). Therefore we are also facing the same problem mentioned in this question, do you have any suggestion how to do achieve it ? Or do you think it makes sense if we open a PR and work on it with your guidance ?

Thanks
Ilter

**Comment by aler9** [comment]
Hello, in order to design a system that stores streams to disk, let's start from choosing the container: a standard MP4 container is generally not the best choice for storing live video streams, since a MP4 works by generating a table of contents (that is exactly like the table of contents of a magazine, it contains a list of all things included in the magazine, in this case video samples, and their location) and placing it the end of the file, only after all samples have been written. If your server crashes, this ToC is not written, and the MP4 file becomes completely useless. In order to store live streams in a reliable way, you have to use a container that is resilient to system crashes. There are mainly two choices: MPEG-TS (.ts) that doesn't use the ToC or fragmented MP4 (fMP4) that use multiple, small, ToCs.

Second, let's choose a muxer, i.e. the software that will generate the container format. In my opinion, using ffmpeg into runOnReady is the best choice, you don't have to hack the code and you can use it directly to save streams to files, as explained in the README:

```yml
paths:
  all:
  original:
    runOnReady: ffmpeg -i rtsp://localhost:$RTSP_PORT/$RTSP_PATH -c copy -f segment -strftime 1 -segment_time 60 -segment_format mpegts saved_%Y-%m-%d_%H-%M-%S.ts
    runOnReadyRestart: yes
```

If you need MP4 files, you can write a script that, for instance, takes all .ts files that are older than 5 minutes, and uses ffmpeg to convert them to MP4.

Alternatively, yes, you can edit the server in order to save streams on disk, but you need some expertise about Golang and containers. It's not trivial, but it can be done. You need to hijack into [stream.WriteData()](https://github.com/aler9/rtsp-simple-server/blob/40aa24a6d8f9b8f3cab37f93895829a29c046f0c/internal/core/stream.go#L129) not, rtmp_conn.go, otherwise your code will work only with RTMP publishers and not with RTSP ones.

If you want to use the MPEG-TS format, i suggest you to take a look at [muxer_variant_mpegts_segment.go](https://github.com/aler9/rtsp-simple-server/blob/main/internal/hls/muxer_variant_mpegts_segment.go), it clearly shows how to save H264 and AAC tracks into a .TS file, and it will work.

Regarding fMP4, keep in mind that the server was designed to generate LL-HLS, not plain fMP4 files, and LL-HLS uses fMP4 in a complex way, it splits the fMP4 into an init.mp4 that contains the track list and a series of .mp4 segments that contains a ToC and some samples. In order to generate a working .mp4, you need to write into a file the content of init.mp4 and then the content of each segment. Therefore, you have to create a file, call [mp4InitGenerate()](https://github.com/aler9/rtsp-simple-server/blob/40aa24a6d8f9b8f3cab37f93895829a29c046f0c/internal/hls/muxer_variant_fmp4_init.go#L521) and append its output to the file, then create a queue of samples, and when this queue is full (for instance, it reaches 100 samples) use this queue to call [mp4PartGenerate()](https://github.com/aler9/rtsp-simple-server/blob/40aa24a6d8f9b8f3cab37f93895829a29c046f0c/internal/hls/muxer_variant_fmp4_part.go#L183), and append its output to the file. This must repeated periodically, and you obtain a valid and resilient fMP4 file, that can be read by most players and browsers.

**Reply by zt1983811** [reply]
@aler9 , I followed your advice and tried to understand how to generate TS file [muxer_variant_mpegts_segment.go](https://github.com/aler9/rtsp-simple-server/blob/main/internal/hls/muxer_variant_mpegts_segment.go), It seems it's just matter of calling
```
func (w *Writer) WriteH264(
	pcr time.Duration,
	dts time.Duration,
	pts time.Duration,
	idrPresent bool,
	nalus [][]byte,
)
```
this function. 
In `stream.go` I believe I have the `pts` and `nalus` from data param. What I am missing is pcr and dts, idrPresent
I googled quick quick, If I know the bitrate of video stream and timebase then I can calculate PCR from PTS like following
`PCR = (PTS * 2,000,000) / 90,000` 
also if I know frame rate and GOP size then I can calculate DTS from PCR
`DTS = PCR - (12 / 25)`
I am wondering where I can get this information in `stream.go` ?

**Reply by aler9** [reply]
You can set PCR to `time.Now().Sub(startTime)` and compute DTS by using `h264.DTSExtractor`. There's an example here:

https://github.com/aler9/gortsplib/tree/main/examples/client-read-format-h264-save-to-disk

[a comment by an automated account is not shown]
